api/docs/v2/complex_commands/parameters.rst
:og:description: Parameters for fine-tuning complex liquid handling behavior in the Python API.

.. _complex_params:

**********************************
Complex Liquid Handling Parameters
**********************************

Complex commands accept a number of optional parameters that give you greater control over the exact steps they perform.  

This page describes the accepted values and behavior of each parameter. The parameters are organized in the order that they first add a step. Some parameters, such as ``touch_tip``, add multiple steps. See :ref:`complex-command-order` for more details on the sequence of steps performed by complex commands.

The API reference entry for :py:meth:`.InstrumentContext.transfer` also lists the parameters and has more information on their implementation as keyword arguments.

.. _param-tip-handling:

Tip Handling
============

The ``new_tip`` parameter controls if and when complex commands pick up new tips from the pipette's tip racks. It has three possible values:

.. list-table::
   :header-rows: 1

   * - Value
     - Behavior
   * - ``"once"``
     - 
        - Pick up a tip at the start of the command.
        - Use the tip for all liquid handling.
        - Drop the tip at the end of the command.
   * - ``"always"``
     - Pick up and drop a tip for each set of aspirate and dispense steps.
   * - ``"never"``
     - Do not pick up or drop tips at all.
     
``"once"`` is the default behavior for all complex commands.

.. versionadded:: 2.0
     
Tip Handling Requirements
-------------------------
     
``"once"`` and ``"always"`` require that the pipette has an :ref:`associated tip rack <pipette-tip-racks>`, or the API will raise an error (because it doesn't know where to pick up a tip from). If the pipette already has a tip attached, the API will also raise an error when it tries to pick up a tip. 

.. code-block:: python

    pipette.pick_up_tip()
    pipette.transfer(
        volume=100,
        source=plate["A1"],
        dest=[plate["B1"], plate["B2"], plate["B3"]],
        new_tip="never",  # "once", "always", or None will error
    )

Conversely, ``"never"`` requires that the pipette has picked up a tip, or the API will raise an error (because it will attempt to aspirate without a tip attached).

Avoiding Cross-Contamination
----------------------------

One reason to set ``new_tip="always"`` is to avoid cross-contamination between wells. However, you should always do a dry run of your protocol to test that the pipette is picking up and dropping tips in the way that your application requires.

:py:meth:`~.InstrumentContext.transfer` will pick up a new tip before *every* aspirate when ``new_tip="always"``. This includes when :ref:`tip refilling <complex-tip-refilling>` requires multiple aspirations from a single source well.

:py:meth:`~.InstrumentContext.distribute` and :py:meth:`~.InstrumentContext.consolidate` only pick up one tip, even when ``new_tip="always"``. For example, this distribute command returns to the source well a second time, because the amount to be distributed (400 µL total plus disposal volume) exceeds the pipette capacity (300 µL)::

    pipette.distribute(
        volume=200,
        source=plate["A1"],
        dest=[plate["B1"], plate["B2"]],
        new_tip="always",
    )
    
But it *does not* pick up a new tip after dispensing into B1:

.. code-block:: text

    Picking up tip from A1 of tip rack on 3
    Aspirating 220.0 uL from A1 of well plate on 2 at 92.86 uL/sec
    Dispensing 200.0 uL into B1 of well plate on 2 at 92.86 uL/sec
    Blowing out at A1 of Opentrons Fixed Trash on 12
    Aspirating 220.0 uL from A1 of well plate on 2 at 92.86 uL/sec
    Dispensing 200.0 uL into B2 of well plate on 2 at 92.86 uL/sec
    Blowing out at A1 of Opentrons Fixed Trash on 12
    Dropping tip into A1 of Opentrons Fixed Trash on 12

If this poses a contamination risk, you can work around it in a few ways:

    * Use ``transfer()`` with ``new_tip="always"`` instead.
    * Set :py:obj:`.well_bottom_clearance` high enough that the tip doesn't contact liquid in the destination well.
    * Use :ref:`building block commands <v2-atomic-commands>` instead of complex commands.


.. _param-mix-before:

Mix Before
==========

The ``mix_before`` parameter controls mixing in source wells before each aspiration. Its value must be a :py:class:`tuple` with two numeric values. The first value is the number of repetitions, and the second value is the amount of liquid to mix in µL.

For example, this transfer command will mix 50 µL of liquid 3 times before each of its aspirations::

    pipette.transfer(
        volume=100,
        source=plate["A1"],
        dest=[plate["B1"], plate["B2"]],
        mix_before=(3, 50),
    )
    
.. versionadded:: 2.0

Mixing occurs before every aspiration, including when :ref:`tip refilling <complex-tip-refilling>` is required.

.. note::
    :py:meth:`~.InstrumentContext.consolidate` ignores any value of ``mix_before``. Mixing on the second and subsequent aspirations of a consolidate command would defeat its purpose: to aspirate multiple times in a row, from different wells, *before* dispensing.
    
.. _param-disposal-volume:

Disposal Volume
===============

The ``disposal_volume`` parameter controls how much extra liquid is aspirated as part of a :py:meth:`~.InstrumentContext.distribute` command. Including a disposal volume can improve the accuracy of each dispense. The pipette blows out the disposal volume of liquid after dispensing. To skip aspirating and blowing out extra liquid, set ``disposal_volume=0``.

By default, ``disposal_volume`` is the :ref:`minimum volume <new-pipette-models>` of the pipette, but you can set it to any amount::

    pipette.distribute(
        volume=100,
        source=plate["A1"],
        dest=[plate["B1"], plate["B2"]],
        disposal_volume=10,  # reduce from default 20 µL to 10 µL
    )
    
.. versionadded:: 2.0
    
If the amount to aspirate plus the disposal volume exceeds the tip's capacity, ``distribute()`` will use a :ref:`tip refilling strategy <complex-tip-refilling>`. In such cases, the pipette will aspirate and blow out the disposal volume *for each aspiration*. For example, this command will require tip refilling with a 1000 µL pipette::
    
    pipette.distribute(
        volume=120,
        source=reservoir["A1"],
        dest=[plate.columns()[0]],
        disposal_volume=50,
    )
    
The amount to dispense in the destination is 960 µL (120 µL for each of 8 wells in the column). Adding the 50 µL disposal volume exceeds the 1000 µL capacity of the tip. The command will be split across two aspirations, each with the full disposal volume of 50 µL. The pipette will dispose *a total of 100 µL* during the command.

.. note::
    :py:meth:`~.InstrumentContext.transfer` will not aspirate additional liquid if you set ``disposal_volume``. However, it will perform a very small blow out after each dispense.
    
    :py:meth:`~.InstrumentContext.consolidate` ignores ``disposal_volume`` completely.

.. _param-touch-tip:

Touch Tip
=========

The ``touch_tip`` parameter accepts a Boolean value. When ``True``, a touch tip step occurs after every aspirate and dispense.

For example, this transfer command aspirates, touches the tip at the source, dispenses, and touches the tip at the destination::

    pipette.transfer(
        volume=100,
        dest=plate["A1"],
        source=plate["B1"],
        touch_tip=True,
    )

.. versionadded:: 2.0

Touch tip occurs after every aspiration, including when :ref:`tip refilling <complex-tip-refilling>` is required.

This parameter always uses default motion behavior for touch tip. Use the :ref:`touch tip building block command <touch-tip>` if you need to:

    * Only touch the tip after aspirating or dispensing, but not both.
    * Control the speed, radius, or height of the touch tip motion.

.. _param-air-gap:

Air Gap
=======

The ``air_gap`` parameter controls how much air to aspirate and hold in the bottom of the tip when it contains liquid. The parameter's value is the amount of air to aspirate in µL.

Air-gapping behavior is different for each complex command. The different behaviors all serve the same purpose, which is to never leave the pipette holding liquid at the very bottom of the tip. This helps keep liquids from seeping out of the pipette.

.. list-table::
   :header-rows: 1

   * - Method
     - Air-gapping behavior
   * - ``transfer()``
     - 
       - Air gap after each aspiration.
       - Pipette is empty after dispensing.
   * - ``distribute()``
     - 
       - Air gap after each aspiration.
       - Air gap after dispensing if the pipette isn't empty.
   * - ``consolidate()``
     - 
       - Air gap after each aspiration. This may create multiple air gaps within the tip.
       - Pipette is empty after dispensing.

For example, this transfer command will create a 20 µL air gap after each of its aspirations. When dispensing, it will clear the air gap and dispense the full 100 µL of liquid::

    pipette.transfer(
        volume=100,
        source=plate["A1"],
        dest=plate["B1"],
        air_gap=20,
    )
    
.. versionadded:: 2.0

When consolidating, air gaps still occur after every aspiration. In this example, the tip will use 210 µL of its capacity (50 µL of liquid followed by 20 µL of air, repeated three times)::

    pipette.consolidate(
        volume=50,
        source=[plate["A1"], plate["A2"], plate["A3"]],
        dest=plate["B1"],
        air_gap=20,
    )

.. code-block:: text

    Picking up tip from A1 of tip rack on 3
    Aspirating 50.0 uL from A1 of well plate on 2 at 92.86 uL/sec
    Air gap
        Aspirating 20.0 uL from A1 of well plate on 2 at 92.86 uL/sec
    Aspirating 50.0 uL from A2 of well plate on 2 at 92.86 uL/sec
    Air gap
        Aspirating 20.0 uL from A2 of well plate on 2 at 92.86 uL/sec
    Aspirating 50.0 uL from A3 of well plate on 2 at 92.86 uL/sec
    Air gap
        Aspirating 20.0 uL from A3 of well plate on 2 at 92.86 uL/sec
    Dispensing 210.0 uL into B1 of well plate on 2 at 92.86 uL/sec
    Dropping tip into A1 of Opentrons Fixed Trash on 12
    
If adding an air gap would exceed the pipette's maximum volume, the complex command will use a :ref:`tip refilling strategy <complex-tip-refilling>`. For example, this command uses a 300 µL pipette to transfer 300 µL of liquid plus an air gap::

    pipette.transfer(
        volume=300,
        source=plate["A1"],
        dest=plate["B1"],
        air_gap=20,
    )

As a result, the transfer is split into two aspirates of 150 µL, each with their own 20 µL air gap:

.. code-block:: text

	Picking up tip from A1 of tip rack on 3
	Aspirating 150.0 uL from A1 of well plate on 2 at 92.86 uL/sec
	Air gap
		Aspirating 20.0 uL from A1 of well plate on 2 at 92.86 uL/sec
	Dispensing 170.0 uL into B1 of well plate on 2 at 92.86 uL/sec
	Aspirating 150.0 uL from A1 of well plate on 2 at 92.86 uL/sec
	Air gap
		Aspirating 20.0 uL from A1 of well plate on 2 at 92.86 uL/sec
	Dispensing 170.0 uL into B1 of well plate on 2 at 92.86 uL/sec
	Dropping tip into A1 of Opentrons Fixed Trash on 12

.. _param-mix-after:

Mix After
=========

The ``mix_after`` parameter controls mixing in source wells after each dispense. Its value must be a :py:class:`tuple` with two numeric values. The first value is the number of repetitions, and the second value is the amount of liquid to mix in µL.

For example, this transfer command will mix 50 µL of liquid 3 times after each of its dispenses::

    pipette.transfer(
        volume=100,
        source=plate["A1"],
        dest=[plate["B1"], plate["B2"]],
        mix_after=(3, 50),
    )
    
.. versionadded:: 2.0

.. note::
    :py:meth:`~.InstrumentContext.distribute` ignores any value of ``mix_after``. Mixing after dispensing would combine (and potentially contaminate) the remaining source liquid with liquid present at the destination.

.. _param-blow-out:

Blow Out
========

There are two parameters that control whether and where the pipette blows out liquid. The ``blow_out`` parameter accepts a Boolean value. When ``True``, the pipette blows out remaining liquid when the tip is empty or only contains the disposal volume. The ``blowout_location`` parameter controls in which of three locations these blowout actions occur. The default blowout location is the trash. Blowout behavior is different for each complex command. 

.. list-table::
   :header-rows: 1

   * - Method
     - Blowout behavior and location
   * - ``transfer()``
     -
       - Blow out after each dispense.
       - Valid locations: ``"trash"``, ``"source well"``, ``"destination well"``
   * - ``distribute()``
     - 
       - Blow out after the final dispense.
       - Valid locations: ``"trash"``, ``"source well"``
   * - ``consolidate()``
     - 
       - Blow out after the only dispense.
       - Valid locations: ``"trash"``, ``"destination well"``

For example, this transfer command will blow out liquid in the trash twice, once after each dispense into a destination well::

    pipette.transfer(
        volume=100,
        source=[plate["A1"], plate["A2"]],
        dest=[plate["B1"], plate["B2"]],
        blow_out=True,
    )

.. versionadded:: 2.0

Set ``blowout_location`` when you don't want to waste any liquid by blowing it out into the trash. For example, you may want to make sure that every last bit of a sample is moved into a destination well. Or you may want to return every last bit of an expensive reagent to the source for use in later pipetting. 

If you need to blow out in a different well, or at a specific location within a well, use the :ref:`blow out building block command <blow-out>` instead.

When setting a blowout location, you *must* also set ``blow_out=True``, or the location will be ignored::

    pipette.transfer(
        volume=100,
        source=plate["A1"],
        dest=plate["B1"],
        blow_out=True,  # required to set location
        blowout_location="destination well",
    )

.. versionadded:: 2.8

With ``transfer()``, the pipette will not blow out at all if you only set ``blowout_location``.

``blow_out=True`` is also required for distribute commands that blow out by virtue of having a disposal volume::

    pipette.distribute(
        volume=100,
        source=plate["A1"],
        dest=[plate["B1"], plate["B2"]],
        disposal_volume=50,  # causes blow out
        blow_out=True,       # still required to set location!
        blowout_location="source well",
    )

With ``distribute()``, the pipette will still blow out if you only set ``blowout_location``, but in the default location of the trash.

.. note::
    If the tip already contains liquid before the complex command, the default blowout location will shift away from the trash. ``transfer()`` and ``distribute()`` shift to the source well, and ``consolidate()`` shifts to the destination well. For example, this transfer command will blow out in well B1 because it's the source::
    
        pipette.pick_up_tip()
        pipette.aspirate(100, plate["A1"])    
        pipette.transfer(
            volume=100,
            source=plate["B1"],
            dest=plate["C1"],
            new_tip="never",
            blow_out=True,
            # no blowout_location
        )
        pipette.drop_tip()

    This only occurs when you aspirate and then perform a complex command with ``new_tip="never"`` and ``blow_out=True``.

.. _param-trash:

Trash Tips
==========

The ``trash`` parameter controls what the pipette does with tips at the end of complex commands. When ``True``, the pipette drops tips into the trash. When ``False``, the pipette returns tips to their original locations in their tip rack. 

The default is ``True``, so you only have to set ``trash`` when you want the tip-returning behavior::

    pipette.transfer(
        volume=100,
        source=plate["A1"],
        dest=plate["B1"],
        trash=False,
    )

.. versionadded:: 2.0

api/release-notes.md
For more details about this release, please see the full [technical change
log][]. For a list of currently known issues, please see the [Opentrons issue tracker][].

[technical change log]: https://github.com/Opentrons/opentrons/releases
[opentrons issue tracker]: https://github.com/Opentrons/opentrons/issues?q=is%3Aopen+is%3Aissue+label%3Abug

By installing and using Opentrons software, you agree to the Opentrons End-User License Agreement (EULA). You can view the EULA at [opentrons.com/eula](https://opentrons.com/eula).

---
## Opentrons Robot Software Changes in 8.4.0

Welcome to the v8.4.0 release of the Opentrons robot software! This release includes updated liquid handling commands for Opentrons-verified liquid classes and other new features, improvements, and bug fixes. 

### New Features

- Use new ``transfer_liquid``, ``distribute_liquid``, and ``consolidate_liquid`` commands on Flex to optimize liquid handling based on Opentrons-verified liquid classes. 
- Stack multiple Opentrons Tough Auto-Sealing Lids on the deck. 
- Move Opentrons Tough Auto-Sealing Lids or remove tip rack lids with the Flex Gripper. 
- Aspirate or dispense in a well based on the liquid meniscus. 

### Improvements

- Improved tip drop motions for Flex 1- and 8-channel pipettes correctly drop and return attached tips to the tip rack. 

### Bug Fixes

- The API raises an error if the Absorbance Plate Reader fails to initialize before reading a plate inside the module. 

## Opentrons Robot Software Changes in 8.3.1

The 8.3.1 hotfix release contains two bug fixes:    

- Exported data from the Absorbance Plate Reader no longer contains invalid values. 
- A small fix allows all robots to properly reboot after an upgrade to v8.3.0. 

## Opentrons Robot Software Changes in 8.3.0

Welcome to the v8.3.0 release of the Opentrons robot software! This release includes improvements to error recovery on the Flex, as well as beta features for our commercial partners. 


### Improved Features

- Improvements to the Flex error recovery feature help protocols recover from detected stalls and collisions, saving you valuable time and resources.

---

## Opentrons Robot Software Changes in 8.2.0

Welcome to the v8.2.0 release of the Opentrons robot software! This release adds support for the Opentrons Absorbance Plate Reader Module.

### New Features

- Create and run Python protocols that use the Opentrons Absorbance Plate Reader.

### Improved Features

- Liquid presence detection no longer checks for liquid before every aspiration in a `mix()` command.

### Bug Fixes

- Error recovery no longer causes an `AssertionError` when a Python protocol changes the pipette speed.

### Known Issues

- You can't downgrade the robot software with an Absorbance Plate Reader attached. Disconnect the module first if you need to downgrade.

---

## Opentrons Robot Software Changes in 8.1.0

Welcome to the v8.1.0 release of the Opentrons robot software!

### Hardware Support

- Latest production version of Flex robots

---

## Opentrons Robot Software Changes in 8.0.0

Welcome to the v8.0.0 release of the Opentrons robot software!

### New Features

- Create, store, and run quick transfers on Flex.
- Define and use CSV runtime parameters in Python protocols.
- Detect the presence or absence of liquid in a well (Flex pipettes only), and continue or pause the protocol based on the result.
- Automatically pause Flex protocol runs when detecting overpressure, allowing for error recovery and run resumption.

### Improved Features

- Provides more partial tip pickup configurations. All multi-channel pipettes now support single and partial column pickup, and the Flex 96-channel pipette now supports row pickup.
- Improves homing behavior when a Flex protocol completes or is canceled with liquid-filled tips attached to the pipette.

### Known Issues

- During single-tip or partial-column pickup with a multi-channel pipette, tips in adjacent columns may cling to empty nozzles. Pick up tips row by row, rather than column by column, to avoid this.
- Protocol analysis and `opentrons_simulate` do not raise an error when a protocol tries to detect liquid with a pipette nozzle configuration that doesn't contain a pressure sensor (single-tip pickup with A12 or H1). Avoid using the A12 and H1 nozzles for single-tip pickup if you need to detect liquid presence within wells.
- `opentrons_simulate` describes motion to wells only with respect to the primary channel, regardless of the current pipette nozzle configuration.

---

## Opentrons Robot Software Changes in 7.5.0

Welcome to the v7.5.0 release of the Opentrons robot software!

### Hardware Support

- [Opentrons Flex HEPA/UV Module](https://opentrons.com/products/opentrons-flex-hepa-uv-module)
- Latest Flex Gripper model (serial numbers beginning `GRPV13`)

### Bug Fixes

- Fixed certain string runtime parameter values being misinterpreted as an incorrect type.
- The `opentrons_execute` command-line tool and `opentrons.execute` Python API functions now take the deck configuration of Flex into account when planning gantry movement.

### Known Issue

- The HEPA/UV Module's buttons may not respond properly after its safety shutoff is activated. This happens when the module is removed from the top of Flex while its lights are on. Power cycle the module to restore normal behavior. The module is safe to use even if you do not power cycle it.

---

## Opentrons Robot Software Changes in 7.3.1

Welcome to the v7.3.1 release of the Opentrons robot software!

### Improved Features

- Updated values for how much a tip overlaps with the pipette nozzle when the pipette picks up tips, in order to make protocols more reliable. These new values only apply to JSON protocols and Python protocols specifying API version 2.19.

---

## Opentrons Robot Software Changes in 7.3.0

Welcome to the v7.3.0 release of the Opentrons robot software!

### New Features


hardware-testing/hardware_testing/gravimetric/config.py
"""Config."""
from dataclasses import dataclass
from typing import List, Dict, Tuple
from typing_extensions import Final
from enum import Enum
from opentrons.config.types import LiquidProbeSettings
from opentrons.protocol_api.labware import Well


class ConfigType(Enum):
    """Substitute for Literal which isn't available until 3.8.0."""

    gravimetric = 1
    photometric = 2


@dataclass
class VolumetricConfig:
    """Execute shared volumetric Setup Config."""

    name: str
    pipette_volume: int
    pipette_channels: int
    pipette_mount: str
    tip_volume: int
    trials: int
    slots_tiprack: List[int]
    increment: bool
    return_tip: bool
    mix: bool
    user_volumes: bool
    kind: ConfigType
    extra: bool
    jog: bool
    same_tip: bool
    ignore_fail: bool
    mode: str


@dataclass
class GravimetricConfig(VolumetricConfig):
    """Execute Gravimetric Setup Config."""

    labware_on_scale: str
    slot_scale: int
    blank: bool
    gantry_speed: int
    scale_delay: int
    isolate_channels: List[int]
    isolate_volumes: List[float]
    liquid: str


@dataclass
class PhotometricConfig(VolumetricConfig):
    """Execute photometric Setup Config."""

    photoplate: str
    photoplate_slot: int
    reservoir: str
    reservoir_slot: int
    touch_tip: bool
    refill: bool
    photoplate_column_offset: List[int]
    dye_well_column_offset: List[int]


GRAV_CONFIG_EXCLUDE_FROM_REPORT = ["labware_offsets", "slots_tiprack"]
PHOTO_CONFIG_EXCLUDE_FROM_REPORT = ["labware_offsets", "slots_tiprack"]

NUM_BLANK_TRIALS: Final = 10
NUM_MIXES_BEFORE_ASPIRATE = 5
SCALE_SECONDS_TO_TRUE_STABILIZE = 60 * 3

LOW_VOLUME_UPPER_LIMIT_UL: Final = 2.0

TOUCH_TIP_SPEED = 30

GANTRY_MAX_SPEED = 40
TIP_SPEED_WHILE_SUBMERGING_ASPIRATE = 50
TIP_SPEED_WHILE_SUBMERGING_DISPENSE = 50
TIP_SPEED_WHILE_RETRACTING_ASPIRATE = 50
TIP_SPEED_WHILE_RETRACTING_DISPENSE = 50

VIAL_SAFE_Z_OFFSET: Final = 25
LABWARE_BOTTOM_CLEARANCE = 1.5

LIQUID_PROBE_SETTINGS: Dict[int, Dict[int, Dict[int, Dict[str, int]]]] = {
    50: {
        1: {
            20: {
                "mount_speed": 5,
                "plunger_speed": 15,
                "sensor_threshold_pascals": 15,
            },
            50: {
                "mount_speed": 5,
                "plunger_speed": 15,
                "sensor_threshold_pascals": 15,
            },
        },
        8: {
            50: {
                "mount_speed": 5,
                "plunger_speed": 15,
                "sensor_threshold_pascals": 15,
            },
        },
    },
    200: {
        96: {
            20: {
                "mount_speed": 5,
                "plunger_speed": 5,
                "sensor_threshold_pascals": 15,
            },
            50: {
                "mount_speed": 5,
                "plunger_speed": 5,
                "sensor_threshold_pascals": 15,
            },
            200: {
                "mount_speed": 5,
                "plunger_speed": 5,
                "sensor_threshold_pascals": 15,
            },
        }
    },
    1000: {
        1: {
            50: {
                "mount_speed": 5,
                "plunger_speed": 15,
                "sensor_threshold_pascals": 15,
            },
            200: {
                "mount_speed": 5,
                "plunger_speed": 15,
                "sensor_threshold_pascals": 15,
            },
            1000: {
                "mount_speed": 5,
                "plunger_speed": 15,
                "sensor_threshold_pascals": 15,
            },
        },
        8: {
            50: {
                "mount_speed": 5,
                "plunger_speed": 15,
                "sensor_threshold_pascals": 15,
            },
            200: {
                "mount_speed": 5,
                "plunger_speed": 15,
                "sensor_threshold_pascals": 15,
            },
            1000: {
                "mount_speed": 5,
                "plunger_speed": 15,
                "sensor_threshold_pascals": 15,
            },
        },
        96: {
            20: {
                "mount_speed": 5,
                "plunger_speed": 5,
                "sensor_threshold_pascals": 15,
            },
            50: {
                "mount_speed": 5,
                "plunger_speed": 5,
                "sensor_threshold_pascals": 15,
            },
            200: {
                "mount_speed": 5,
                "plunger_speed": 5,
                "sensor_threshold_pascals": 15,
            },
            1000: {
                "mount_speed": 5,
                "plunger_speed": 5,
                "sensor_threshold_pascals": 15,
            },
        },
    },
}


def _get_liquid_probe_settings(
    cfg: VolumetricConfig, well: Well
) -> LiquidProbeSettings:
    lqid_cfg: Dict[str, int] = LIQUID_PROBE_SETTINGS[cfg.pipette_volume][
        cfg.pipette_channels
    ][cfg.tip_volume]
    return LiquidProbeSettings(
        mount_speed=lqid_cfg["mount_speed"],
        plunger_speed=lqid_cfg["plunger_speed"],
        plunger_impulse_time=0.2,
        sensor_threshold_pascals=lqid_cfg["sensor_threshold_pascals"],
        aspirate_while_sensing=False,
        z_overlap_between_passes_mm=0.1,
        plunger_reset_offset=2.0,
        samples_for_baselining=20,
        sample_time_sec=0.004,
    )


QC_VOLUMES_G: Dict[int, Dict[int, List[Tuple[int, List[float]]]]] = {
    1: {
        50: [  # P50
            (20, [1.0, 20.0]),
            (50, [1.0, 50.0]),  # T50
        ],
        1000: [  # P1000
            (50, [5.0]),  # T50
            (200, []),  # T200
            (1000, [1000.0]),  # T1000
        ],
    },
    8: {
        50: [  # P50
            (50, [1.0, 50.0]),  # T50
        ],
        1000: [  # P1000
            (50, [5.0]),  # T50
            (200, []),  # T200
            (1000, [1000.0]),  # T1000
        ],
    },
    96: {
        200: [
            (20, [0.5, 1.0]),  # T20
            (50, [1.0, 50.0]),  # T50
            (200, [200.0]),  # T200
        ],
        1000: [  # P1000
            (20, [5.0]),
            (50, [5.0]),  # T50
            (200, [200.0]),  # T200
            (1000, [1000.0]),  # T1000
        ],
    },
}


QC_VOLUMES_EXTRA_G: Dict[int, Dict[int, List[Tuple[int, List[float]]]]] = {
    1: {
        50: [  # P50
            (50, [10.0]),  # T50
        ],
        1000: [  # P1000
            (50, [50]),  # T50
            (200, [200.0]),  # T200
            (1000, []),  # T1000
        ],
    },
    8: {
        50: [  # P50
            (50, [10.0]),  # T50
        ],
        1000: [  # P1000
            (50, [50.0]),  # T50
            (200, [200.0]),  # T200
            (1000, []),  # T1000
        ],
    },
    96: {
        1000: [  # P1000
            (50, []),  # T50
            (200, []),  # T200
            (1000, []),  # T1000
        ],
    },
}

QC_VOLUMES_P: Dict[int, Dict[int, List[Tuple[int, List[float]]]]] = {
    1: {
        50: [  # P50
            (50, [1.0]),
        ],
        1000: [  # P1000
            (50, [5.0]),  # T50
            (200, [200.0]),  # T200
            (1000, []),  # T1000
        ],
    },
    8: {
        50: [  # P50
            (50, [1.0]),
        ],
        1000: [  # P1000
            (50, [5.0]),  # T50
            (200, [200.0]),  # T200
            (1000, []),  # T1000
        ],
    },
    96: {
        200: [
            (20, [1.0, 0.5]),  # T50
            (50, [50.0]),  # T50
            (200, [200.0]),  # T200
        ],
        1000: [  # P1000
            (50, [5.0]),  # T50
            (200, [200.0]),  # T200
            (1000, []),  # T1000
        ],
    },
}

QC_DEFAULT_TRIALS: Dict[ConfigType, Dict[int, int]] = {
    ConfigType.gravimetric: {
        1: 10,
        8: 10,
        96: 9,
    },
    ConfigType.photometric: {
        1: 8,
        8: 12,
        96: 5,
    },
}

QC_TEST_SAFETY_FACTOR = 0.0

QC_TEST_MIN_REQUIREMENTS: Dict[
    int, Dict[int, Dict[int, Dict[float, Tuple[float, float]]]]
] = {
    # channels: [Pipette: [tip: [Volume: (%d, Cv)]]]
    1: {
        50: {  # P50
            20: {
                1.0: (5.0, 4.0),
                10.0: (1.0, 0.5),
                20.0: (1, 0.4),
            },
            50: {
                1.0: (5.0, 4.0),
                10.0: (1.0, 0.5),
                50.0: (1, 0.4),
            },
        },  # T50
        1000: {  # P1000
            50: {  # T50
                5.0: (5.0, 5.0),
                10.0: (2.0, 2.0),
                50.0: (1.0, 1.0),
            },
            200: {  # T200
                5.0: (7.0, 4.00),
                50.0: (2.0, 1.0),
                200.0: (0.5, 0.2),
            },
            1000: {  # T1000
                10.0: (7.5, 3.5),
                100.0: (2.0, 0.75),
                1000.0: (0.7, 0.15),
            },
        },
    },
    8: {
        50: {  # P50
            50: {  # T50
                1.0: (20.0, 5.0),
                10.0: (3.0, 2.0),
                50.0: (1.25, 0.4),
            },
        },
        1000: {  # P1000
            50: {  # T50
                5.0: (5.0, 5.0),
                10.0: (1.5, 1.5),
                50.0: (1.0, 1.0),
            },
            200: {  # T200
                5.0: (5.0, 5.0),
                50.0: (1.5, 1.5),
                200.0: (1.0, 0.4),
            },
            1000: {  # T1000
                10.0: (10.0, 5.0),
                100.0: (2.5, 1.0),
                1000.0: (0.7, 0.15),
            },
        },
    },
    96: {
        200: {
            20: {  # T20
                0.5: (2.5, 2.0),
                1.0: (2.5, 2.0),
                2.0: (2.5, 2.0),
                3.0: (2.5, 2.0),
                5.0: (2.5, 2.0),
                10.0: (3.1, 1.7),
            },
            50: {  # T50
                1.0: (2.5, 2.0),
                50.0: (1.5, 0.75),
            },
            200: {  # T200
                5.0: (2.5, 4.0),
                50.0: (1.5, 2.0),
                200.0: (1.4, 0.9),
            },
        },
        1000: {  # P1000
            20: {  # T20
                1.0: (2.5, 2.0),
                2.0: (2.5, 2.0),
                3.0: (2.5, 2.0),
                5.0: (2.5, 2.0),
                10.0: (3.1, 1.7),
                20.0: (3.1, 1.7),
            },
            50: {  # T50
                1.0: (2.5, 2.0),
                2.0: (2.5, 2.0),
                3.0: (2.5, 2.0),
                5.0: (2.5, 2.0),
                10.0: (3.1, 1.7),
                50.0: (1.5, 0.75),
            },
            200: {  # T200
                5.0: (2.5, 4.0),
                50.0: (1.5, 2.0),
                200.0: (1.4, 0.9),
            },
            1000: {  # T1000
                10.0: (5.0, 5.0),
                100.0: (2.5, 1.5),
                1000.0: (1.0, 0.75),
            },
        },
    },
}


def get_tip_volumes_for_qc(
    pipette_volume: int, pipette_channels: int, extra: bool, photometric: bool
) -> List[int]:
    """Build the default testing volumes for qc."""
    config: Dict[int, Dict[int, List[Tuple[int, List[float]]]]] = {}
    tip_volumes: List[int] = []
    if photometric:
        config = QC_VOLUMES_P
    else:
        config = QC_VOLUMES_G
    for t, vls in config[pipette_channels][pipette_volume]:
        if len(vls) > 0 and t not in tip_volumes:
            tip_volumes.append(t)
    if extra:
        for t, vls in QC_VOLUMES_EXTRA_G[pipette_channels][pipette_volume]:
            if len(vls) > 0 and t not in tip_volumes:
                tip_volumes.append(t)

    assert len(tip_volumes) > 0
    return tip_volumes
